src/t_hash.c

```diff
@@ -943,30 +943,25 @@ int hashTypeSet(redisDb *db, robj *o, sds field, sds value, int flags) {
             hashTypeConvert(o, OBJ_ENCODING_HT, &db->hexpires);
 
     } else if (o->encoding == OBJ_ENCODING_HT) {
-        hfield newField = hfieldNew(field, sdslen(field), 0);
         dict *ht = o->ptr;
         dictEntry *de, *existing;
 
-        /* stored key is different than lookup key */
-        dictUseStoredKeyApi(ht, 1);
-        de = dictAddRaw(ht, newField, &existing);
-        dictUseStoredKeyApi(ht, 0);
-
-        /* If field already exists, then update "field". "Value" will be set afterward */
-        if (de == NULL) {
-            if (flags & HASH_SET_KEEP_TTL) {
-                /* keep old field along with TTL */
-                hfieldFree(newField);
-            } else {
+        /* check if field already exists */
+        existing = dictFind(ht, field);
+        /* check if field already exists */
+        if (existing == NULL) {
+            hfield newField = hfieldNew(field, sdslen(field), 0);
+            de = dictAddRaw(ht, newField, NULL);
+        } else {
+            de = existing;
+            /* If field already exists, then update "field". "Value" will be set afterward */
+            if (!(flags & HASH_SET_KEEP_TTL)) {
                 /* If attached TTL to the old field, then remove it from hash's private ebuckets */
                 hfield oldField = dictGetKey(existing);
                 hfieldPersist(o, oldField);
-                hfieldFree(oldField);
-                dictSetKey(ht, existing, newField);
+                dictSetKey(ht, existing, oldField);
+                update = 1;
             }
```

Comment:
The flag `HASH_SET_KEEP_TTL` should have been `HASH_SET_KEEP_TTL_AND_FIELD`.  Otherwise, if flag not set, then we need to:
1.  `hfieldPersist()` old field in order to remove it from TTL ebuckets. 
2. And then replace old field with new field - This is something you overlooked in your code.

Please also be aware to the tradeoffs. In the new code, if `dictFind(ht, field)` return NULL then it will make another dict lookup by `dictAddRaw()`

Reply:
> Please also be aware to the tradeoffs. In the new code, if dictFind(ht, field) return NULL then it will make another dict lookup by dictAddRaw()

@moticless concerning the double lookup, with the introduced API (given we know that the entry does not exist) we can avoid all the overhead. 

WRT:

> And then replace old field with new field - This is something you overlooked in your code.

Why do we need to replace the old field, when the we can just updated the TTL of the old field and avoid replacing. Am I missing something?

Reply:
This function can set new field with TTL or update existing value with TTL. Note that it can be called from: `RM_HashSet()`, `hsetnxCommand()`, `hsetCommand()`, `hincrbyCommand()` and `hincrbyfloatCommand()`.